src/vsr/grid.zig

```diff
@@ -571,117 +661,112 @@ pub fn GridType(comptime Storage: type) type {
 
             const header = schema.header_from_block(cache_block);
             assert(header.op == address);
+            assert(header.cluster == grid.superblock.working.cluster);
 
             if (header.checksum == checksum) {
-                if (constants.verify) grid.verify_read(address, cache_block);
+                if (constants.verify and options.coherent) {
+                    grid.verify_read(address, cache_block);
+                }
 
-                grid.cache_coherent.set(cache_index);
                 return cache_block;
             } else {
-                assert(!grid.cache_coherent.isSet(cache_index));
                 return null;
             }
         }
 
-        pub fn read_block_from_cache_or_storage(
+        pub fn read_block(
             grid: *Grid,
-            callback: *const fn (*Grid.Read, BlockPtrConst) void,
+            callback: ReadBlockCallback,
             read: *Grid.Read,
             address: u64,
             checksum: u128,
-            block_type: BlockType,
+            options: struct {
+                cache_read: bool,
+                cache_write: bool,
+            },
         ) void {
             assert(grid.superblock.opened);
             assert(grid.canceling == null);
-            assert(!grid.superblock.free_set.is_free(address));
-            assert(grid.writing(address, null) != .init);
-
             assert(address > 0);
-            assert(block_type != .reserved);
+
+            switch (callback) {
+                .from_local_storage => {
+                    maybe(grid.checkpointing == null);
+                    // We try to read the block even when it is free — if we recently released it,
+                    // it might be found on disk anyway.
+                    maybe(grid.superblock.free_set.is_free(address));
+                    maybe(grid.writing(address, null) == .create);
+                },
+                .from_local_or_global_storage => {
+                    assert(grid.checkpointing == null);
+                    assert(!grid.superblock.free_set.is_free(address));
+                    assert(grid.writing(address, null) != .create);
+                },
+            }
 
             read.* = .{
                 .callback = callback,
                 .address = address,
                 .checksum = checksum,
-                .block_type = block_type,
-                .read_cache = true,
+                .coherent = callback == .from_local_or_global_storage,
+                .cache_read = options.cache_read,
+                .cache_write = options.cache_write,
```

Comment:
Is it true that `if (cache_write) assert(coherent)` ?

Reply:
In practice yes, but it isn't strictly necessary.